File: libs/js/test/parser.test.ts
import * as dashql_parser from '../';

var parser: dashql_parser.DashQLParser;

beforeAll(async () => {
    parser = await dashql_parser.DashQLParser.create();
});

describe('Parser', () => {
    describe('foo', () => {
        test('synatx error', () => {
            const result = parser.parse("select 1e-04");
            expect(result.root.errorsLength()).toBe(0);
            expect(result.root.statements()!.entriesLength()).toBe(1);
        });
    });
//    describe('errors', () => {
//        test('synatx error', async () => {
//            let result = await parser.parse("?");
//            expect(result.root.errorsLength()).toBe(1);
//            expect(result.root.statementsLength()).toBe(0);
//        });
//
//        test('error recovery', async () => {
//            let result = await parser.parse(`
//                ?select * from foo;
//            `);
//            let program = result.root;
//            expect(program.errorsLength()).toBe(1);
//            expect(program.statementsLength()).toBe(1);
//        });
//    });
//
//    describe('single statements', () => {
//        test('integer parameter', async () => {
//            let result = await parser.parse(`
//                declare parameter days type integer;
//            `);
//            let program = result.root;
//            expect(program.errorsLength()).toBe(0);
//            expect(program.statementsLength()).toBe(1);
//        });
//
//        test('http load', async () => {
//            let result = await parser.parse(`
//                load raw_data from http (
//                    url = 'http://www.google.com',
//                    method = get
//                );
//            `);
//            let program = result.root;
//            expect(program.errorsLength()).toBe(0);
//            expect(program.statementsLength()).toBe(1);
//        });
//
//        test('extract json', async () => {
//            let result = await parser.parse(`
//                extract weather_data from raw_data using json ();
//            `);
//            let program = result.root;
//            expect(program.errorsLength()).toBe(0);
//            expect(program.statementsLength()).toBe(1);
//        });
//
//        test('select 1', async () => {
//            let result = await parser.parse(`
//                select 1;
//            `);
//            let program = result.root;
//            expect(program.errorsLength()).toBe(0);
//            expect(program.statementsLength()).toBe(1);
//        });
//
//        test('query as select 1', async () => {
//            let result = await parser.parse(`
//                query "foo" as select 1;
//            `);
//            let program = result.root;
//            expect(program.errorsLength()).toBe(0);
//            expect(program.statementsLength()).toBe(1);
//        });
//    });
});

File: libs/cpp/test/parser_test.cc
// Copyright (c) 2020 The DashQL Authors

#include "dashql/parser/parser_driver.h"
#include "dashql/parser/yaml_encoder.h"
#include "gtest/gtest.h"
#include "flatbuffers/flatbuffers.h"

using namespace dashql::parser;
using namespace std;

namespace {



TEST(ParserTest, SELECT_FCONST) {
    auto in = R"RAW(
    select 1e-04;
)RAW";
    flatbuffers::FlatBufferBuilder builder;
    auto mo = ParserDriver::Parse(builder, in);
    builder.Finish(mo);
    auto m = flatbuffers::GetRoot<sx::Module>(builder.GetBufferPointer());
    ASSERT_EQ(m->errors()->size(), 0) << m->errors()->Get(0)->message()->c_str();
    ASSERT_EQ(m->statements()->entries()->size(), 1);

    auto s = encodeYAML(*m);
    ASSERT_EQ(s, "");
}

// TEST(ParserTest, ParameterDeclaration) {
//     auto in = R"RAW(
//     declare parameter days type integer;
// )RAW";
//     ParseContext ctx;
//     ctx.Parse(in);
//     ASSERT_EQ(ctx.statements().size(), 1);
//     ASSERT_EQ(ctx.errors().size(), 0);
// }
// 
// TEST(ParserTest, LoadHTTP) {
//     auto in = R"RAW(
//     load raw_data from http (
//         url = 'http://www.google.com',
//         method = get
//     );
// )RAW";
//     ParseContext ctx;
//     ctx.Parse(in);
//     ASSERT_EQ(ctx.statements().size(), 1);
//     ASSERT_EQ(ctx.errors().size(), 0);
// }
// 
// TEST(ParserTest, ExtractJsonPath) {
//     auto in = R"RAW(
//     extract weather_data from raw_data using json ();
// )RAW";
//     ParseContext ctx;
//     ctx.Parse(in);
//     ASSERT_EQ(ctx.statements().size(), 1);
//     ASSERT_EQ(ctx.errors().size(), 0);
// }
// 
// TEST(ParserTest, Query1) {
//     auto in = R"RAW(
//     select 1;
// )RAW";
//     ParseContext ctx;
//     ctx.Parse(in);
//     ASSERT_EQ(ctx.statements().size(), 1);
//     ASSERT_EQ(ctx.errors().size(), 0);
// }
// 
// TEST(ParserTest, Query2) {
//     auto in = R"RAW(
//     query "foo" as select 1;
// )RAW";
//     ParseContext ctx;
//     ctx.Parse(in);
//     ASSERT_EQ(ctx.statements().size(), 1);
//     ASSERT_EQ(ctx.errors().size(), 0);
// }
// 
// TEST(ParserTest, SyntaxError) {
//     auto in = "?";
//     ParseContext ctx;
//     ctx.Parse(in);
//     ASSERT_EQ(ctx.statements().size(), 0);
//     ASSERT_EQ(ctx.errors().size(), 1);
// }
// 
// TEST(ParserTest, SyntaxErrorRecovery) {
//     auto in = "?select * from foo;";
//     ParseContext ctx;
//     ctx.Parse(in);
//     ASSERT_EQ(ctx.statements().size(), 1);
//     ASSERT_EQ(ctx.errors().size(), 1);
// }

}  // namespace
